### PR TITLE
Technical guidelines graphql

### PR DESCRIPTION
**Pull request by akaplya**
## Purpose of this PR

Update GraphQL technical vision

### CODE REVIEW

**Comment by devops-devdocs** [comment]
An admin must run tests on this PR before it can be merged.

File: guides/v2.1/coding-standards/technical-vision/webapi.md [hunk 1]
```diff
@@ -88,13 +88,27 @@ Any new design related to Web API must satisfy the following constraints to keep
...
+1. For modularity purposes, GraphQL configuration must be declared in a separate module. 
+For example, to expose GraphQL for the module `MyModule`, you must create the `schema.graphqls` file in the `MyModuleGraphQl` module.
+1. GraphQL is primarily designed for store-front one-page apps and mobile applications.
+It supports token and cookie authentication, as well as guest access to public queries
+1. All queries must return the 200 HTTP status code.
+If an error occurs, return the error in the response body.
+A 500 status code is allowed when an exception occurs when generating a schema, but not during requests.
 1. The Store code should be passed via headers.
+1. GraphQL types and interfaces SHOULD NOT expose internal IDs (like database identity fields).
+1. Resolvers SHOULD NOT rely on execution resultы of another resolvers except data declared by schema.
```
Comment (roma-glushko) [inline guides/v2.1/coding-standards/technical-vision/webapi.md]: @akaplya , resultы -> results

**Comment by devops-devdocs** [comment]
An admin must run tests on this PR before it can be merged.

File: guides/v2.1/coding-standards/technical-vision/webapi.md [hunk 2]
```diff
@@ -88,13 +88,27 @@ Any new design related to Web API must satisfy the following constraints to keep
...
+1. For modularity purposes, GraphQL configuration must be declared in a separate module. 
+For example, to expose GraphQL for the module `MyModule`, you must create the `schema.graphqls` file in the `MyModuleGraphQl` module.
+1. GraphQL is primarily designed for store-front one-page apps and mobile applications.
+It supports token and cookie authentication, as well as guest access to public queries
+1. All queries must return the 200 HTTP status code.
+If an error occurs, return the error in the response body.
+A 500 status code is allowed when an exception occurs when generating a schema, but not during requests.
 1. The Store code should be passed via headers.
+1. GraphQL types and interfaces SHOULD NOT expose internal IDs (like database identity fields).
+1. Resolvers SHOULD NOT rely on execution results of another resolvers except data declared by schema.
```
Comment (keharper) [inline guides/v2.1/coding-standards/technical-vision/webapi.md]:
```suggestion
1. Resolvers SHOULD NOT rely on execution results of another resolver, except data declared by schema.
```

File: guides/v2.1/coding-standards/technical-vision/webapi.md [hunk 3]
```diff
@@ -88,13 +88,27 @@ Any new design related to Web API must satisfy the following constraints to keep
...
+    Schema is the only contract between resolvers of different levels.
```
Comment (keharper) [inline guides/v2.1/coding-standards/technical-vision/webapi.md]:
```suggestion
    The schema is the only contract between resolvers of different levels.
```

File: guides/v2.1/coding-standards/technical-vision/webapi.md [hunk 4]
```diff
@@ -88,13 +88,27 @@ Any new design related to Web API must satisfy the following constraints to keep
...
+    Resolvers SHOULD NOT share state between themselves except allowed context.
+1. Query resolvers SHOULD support deferred execution to avoid the 
+    [N+1 problem](http://webonyx.github.io/graphql-php/data-fetching/#solving-n1-problem)
+    unless this contradicts domain logic.
+1. Mutations SHOULD be designed in manner to support batch mode unless this contradicts domain logic.
```
Comment (keharper) [inline guides/v2.1/coding-standards/technical-vision/webapi.md]:
```suggestion
1. Mutations SHOULD be designed to support batch mode, unless this contradicts domain logic.
```

File: guides/v2.1/coding-standards/technical-vision/webapi.md [hunk 5]
```diff
@@ -88,13 +88,27 @@ Any new design related to Web API must satisfy the following constraints to keep
...
+1. All queries must return the 200 HTTP status code.
+If an error occurs, return the error in the response body.
+A 500 status code is allowed when an exception occurs when generating a schema, but not during requests.
 1. The Store code should be passed via headers.
+1. GraphQL types and interfaces SHOULD NOT expose internal IDs (like database identity fields).
+1. Resolvers SHOULD NOT rely on execution results of another resolvers except data declared by schema.
+    Schema is the only contract between resolvers of different levels.
+    Resolvers SHOULD NOT share state between themselves except allowed context.
+1. Query resolvers SHOULD support deferred execution to avoid the 
+    [N+1 problem](http://webonyx.github.io/graphql-php/data-fetching/#solving-n1-problem)
```
Comment (keharper) [inline guides/v2.1/coding-standards/technical-vision/webapi.md]:
```suggestion
    [N+1 problem](http://webonyx.github.io/graphql-php/data-fetching/#solving-n1-problem),
```

File: guides/v2.2/coding-standards/technical-guidelines.md [hunk 6]
```diff
@@ -752,9 +752,7 @@ class View extends Template
 
 ## 12. Web API
 
-12.1. Both REST and SOAP API's MUST be exposed.
-
-12.2. All {% glossarytooltip 377dc0a3-b8a7-4dfa-808e-2de37e4c0029 %}Web API{% endglossarytooltip %} GET endpoints MUST return lists of entities.
+[web API technical vision]({{ page.baseurl }}/coding-standards/technical-vision/webapi.html#model-consistency-constraints)
```
Comment (keharper) [inline guides/v2.2/coding-standards/technical-guidelines.md]: You need a full sentence here.

**Comment by dobooth** [comment]
Hi @akaplya, we stopped supporting v2.1, so any PRs that reference that version needs to be modified. Please update this PR to move the change from 2.1  to 2.2. Thanks!

File: guides/v2.1/coding-standards/technical-vision/webapi.md [hunk 7]
```diff
@@ -88,13 +88,27 @@ Any new design related to Web API must satisfy the following constraints to keep
 
 **GraphQL**
 
-1. Unlimited nesting should be supported during requests for related entities. (For example, get Order => Order Items => Products => Related Products)
-1. Field filtration must be performed with SQL queries. Do not filter on the application layer after you've fetched all possible fields.
+1. Unlimited nesting should be supported during requests for related entities.
+(For example, get Order => Order Items => Products => Related Products) 
+1. Field filtration MUST be performed by the corresponding level.
```
Comment (buskamuza) [inline guides/v2.1/coding-standards/technical-vision/webapi.md]: "by the corresponding level" - it's not clear how to use this information. What is "corresponding level"?

File: guides/v2.1/coding-standards/technical-vision/webapi.md [hunk 8]
```diff
@@ -88,13 +88,27 @@ Any new design related to Web API must satisfy the following constraints to keep
...
-1. All queries must return the 200 HTTP status code. If an error occurs, return the error in the response body. A 500 status code is allowed when an exception occurs when generating a schema, but not during requests.
+1. For modularity purposes, GraphQL configuration must be declared in a separate module. 
+For example, to expose GraphQL for the module `MyModule`, you must create the `schema.graphqls` file in the `MyModuleGraphQl` module.
+1. GraphQL is primarily designed for store-front one-page apps and mobile applications.
+It supports token and cookie authentication, as well as guest access to public queries
+1. All queries must return the 200 HTTP status code.
+If an error occurs, return the error in the response body.
+A 500 status code is allowed when an exception occurs when generating a schema, but not during requests.
 1. The Store code should be passed via headers.
+1. GraphQL types and interfaces SHOULD NOT expose internal IDs (like database identity fields).
```
Comment (buskamuza) [inline guides/v2.1/coding-standards/technical-vision/webapi.md]: Maybe it would be good to add what should be used instead? It's hard to follow rules that only prohibit things.
Maybe explaining on examples, such as using SKUs for products. Or explaining what's the recommendation for building entities, so that it's possible to identify it.
Also, it would be good to add some short reasoning to this rule.

**Comment by keharper** [comment]
This was opened against 2.1, which no longer exists. @akaplya please open a new PR.

**Comment by dshevtsov** [comment]
@akaplya GraphQL has been introduced to Magento since 2.3.
Please consider moving the guidelines to 2.3 topics only.
